picard/tagger.py

```diff
@@ -346,105 +363,152 @@ def move_files(self, files, target):
+                    {'directory': parent, 'count': number_of_dirs}
+                )
+            else:
+                self.window.set_statusbar_message(
+                    N_("Adding %(count)d directories ..."),
+                    {'directory': parent, 'count': number_of_dirs}
+                )
+        else:
+            self.window.set_statusbar_message(
+                N_("Adding directory: '%(directory)s' ..."),
+                {'directory': dir_list[0]}
+            )
+
+        list_of_files = []
+        recursively = config.setting['recursively_add_files']
+        for directory in dir_list:
+            if recursively:
+                list_of_files.extend(self._add_directory_recursive(directory))
+            else:
+                list_of_files.extend(self._add_directory_non_recursive(directory))
+        self._add_files(list_of_files)
+
     def add_directory(self, path):
         if config.setting['recursively_add_files']:
-            self._add_directory_recursive(path)
+            list_of_files = self._add_directory_recursive(path)
         else:
-            self._add_directory_non_recursive(path)
+            list_of_files = self._add_directory_non_recursive(path)
+        self._add_files(list_of_files)
 
     def _add_directory_recursive(self, path):
         ignore_hidden = config.setting["ignore_hidden_files"]
-        walk = os.walk(unicode(path))
-
-        def get_files():
-            try:
-                root, dirs, files = next(walk)
-                if ignore_hidden:
-                    dirs[:] = [d for d in dirs if not is_hidden(os.path.join(root, d))]
-            except StopIteration:
-                return None
-            else:
-                number_of_files = len(files)
-                if number_of_files:
-                    mparms = {
-                        'count': number_of_files,
-                        'directory': root,
-                    }
-                    log.debug("Adding %(count)d files from '%(directory)r'" %
-                              mparms)
-                    self.window.set_statusbar_message(
-                        ungettext(
-                            "Adding %(count)d file from '%(directory)s' ...",
-                            "Adding %(count)d files from '%(directory)s' ...",
-                            number_of_files),
-                        mparms,
-                        translate=None,
-                        echo=None
-                    )
-                return (os.path.join(root, f) for f in files)
-
-        def process(result=None, error=None):
-            if result:
-                if error is None:
-                    self.add_files(result)
-                thread.run_task(get_files, process)
-
-        process(True, False)
+        list_to_add = []
+        # walk topdown to allow modifying the sub-dirs to remove hidden dirs
+        for root, dirs, files in os.walk(unicode(path), topdown=True):
```

Comment:
Just a note: I just noticed `os.walk()` has no followlinks parameter, and therefore defaults to False, meaning it will not follow symlinked directories (on systems supporting them). I did a fast test, and verified it. Of course, enabling followlinks would imply to handle infinite loops.